### PR TITLE
feat: integrate payload store in BeaconEngine

### PR DESCRIPTION
**Pull request by Rjected**
An attempt at integrating #2143 to return payload IDs on `engine_forkchoiceUpdated`, and wire up the response in `engine_getPayload`.

This adds a few error variants to `BeaconEngineError`, some need to be converted to the correct error according to [the spec](https://github.com/ethereum/execution-apis/blob/main/src/engine/common.md#errors).

### CODE REVIEW

**Comment by gakonst** [comment]
Does this work interfere with the builder ideas from the issue?

Seems not, and this is just the vanilla tx selection logic that should always exist on every client that supports being used as the EL for a proposer?

**Comment by Rjected** [comment]
> Does this work interfere with the builder ideas from the issue?
>
> Seems not, and this is just the vanilla tx selection logic that should always exist on every client that supports being used as the EL for a proposer?

Yeah exactly, the builder mvp here would be with vanilla selection logic

**Comment by gakonst** [comment]
Cool. Cc @rkrasiuk in case this feels obviously wrong (no need to spend any time here as Matt/Dan are driving this for Hive)

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -188,6 +209,66 @@ where
...
+    }
+
+    /// Called to receive the execution payload associated with a payload build process.
+    pub fn on_get_payload(&self, payload_id: H64) -> Result<ExecutionPayload, BeaconEngineError> {
+        // TODO: Client software SHOULD stop the updating process when either a call to
+        // engine_getPayload with the build process's payloadId is made or SECONDS_PER_SLOT (12s in
+        // the Mainnet configuration) have passed since the point in time identified by the
+        // timestamp parameter.
+
+        // for now just return the output from the payload store
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:265]: these are useful comments.

One thing I'm really happy with is the delegation that's happening 
RPC -> beacon engine -> payload handler

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -188,6 +209,66 @@ where
...
+        // if the payload is deemed VALID and the build process has begun.
+        Ok(ForkchoiceUpdated::new(PayloadStatus::new(
+            PayloadStatusEnum::Valid,
+            Some(state.head_block_hash),
+        ))
+        .with_payload_id(payload_id))
+    }
+
+    /// Called to receive the execution payload associated with a payload build process.
+    pub fn on_get_payload(&self, payload_id: H64) -> Result<ExecutionPayload, BeaconEngineError> {
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:259]: I was wondering if we need that here at all, I guess we don't need the Consensus engine here so I'd lean toward implementing this in he rpc handler directly

File: crates/rpc/rpc-types/src/eth/engine/payload.rs [hunk 3]
```diff
@@ -9,7 +9,7 @@ use serde::{ser::SerializeMap, Deserialize, Serialize, Serializer};
 
 /// And 8-byte identifier for an execution payload.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
-pub struct PayloadId(H64);
+pub struct PayloadId(pub H64);
```
Comment (mattsse) [inline crates/rpc/rpc-types/src/eth/engine/payload.rs:12]: please leave this private
we shouldn't need this after

https://github.com/paradigmxyz/reth/pull/2149

File: crates/consensus/beacon/src/engine/error.rs [hunk 4]
```diff
@@ -1,26 +1,45 @@
...
     PipelineChannelClosed,
+    /// Unknown payload
+    #[error("Unknown payload")]
+    UnknownPayload,
     /// Encountered a payload error.
     #[error(transparent)]
     Payload(#[from] PayloadError),
+    /// Encountered an error during the payload building process.
+    #[error(transparent)]
+    BuilderError(#[from] BuilderError),
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/error.rs:31]: naming is still not great,
I want to rename `reth_miner::error::PayloadError` perhaps `PayloadBuilderError`

**Review by mattsse — APPROVED** [review]
good next steps.

I think I want to change how the RPC layer handles these requests, because delegating messages to the consensus engine is very confusing, will be problematic if we need to make PayloadStore functions async.

For now this is okay because it's just how we handle things atm.

File: crates/consensus/beacon/src/engine/mod.rs (hunk 2 above)
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs]: makes sense, will merge for now but we should definitely move `get_payload` logic to the rpc handler

**Pull request merged**